src/main/holochainManager.ts

```diff
@@ -553,60 +553,55 @@ export class HolochainManager {
+      hashes[entry.entryName] = hash;
+    });
+
+    this.integrityChecker.storeToSignedJSON(path.join(uiDir, 'hashes.json'), hashes);
+    zip.extractAllTo(path.join(uiDir, 'assets'));
+  }
+
+  private storeIcon(uiDir: string, icon: Uint8Array): void {
+    const iconPath = path.join(uiDir, 'icon');
+    fs.writeFileSync(iconPath, icon);
+  }
+
   /**
-   * Extracts and stores ui.zip bytes to a directory named after the sha256 of the bytes
-   * if that directory does not exist yet.
-   * If an icon is passed, the icon gets stored alongside the UI assets.
-   * If no icon is passed but an icon.png file is present in the UI assets,
-   * this icon gets stored instead.
+   * Extracts and stores ui.zip bytes to a directory named after the sha256 hash of the bytes
+   * if that directory does not already exist.
+   * If an icon is provided, it is stored alongside the UI assets.
+   * If no icon is provided but an icon.png file is present in the UI assets,
+   * that icon is stored instead.
    *
-   * @param uiBytes
-   * @returns
+   * @param uiBytes - The bytes of the UI zip file.
+   * @param icon - Optional icon to be stored with the UI assets.
+   * @returns The sha256 hash of the UI zip bytes.
    */
-  storeUiIfNecessary(uiBytes: Array<number>, icon?: string): string {
-    if (icon && typeof icon !== 'string') throw new Error('Icon must of type string.');
+  storeUiIfNecessary(uiBytes: Array<number>, icon?: Uint8Array): string {
+    if (icon && !(icon instanceof Uint8Array)) throw new Error('Icon must be of type Uint8Array.');
 
     // compute UI hash
     const uiZipHasher = crypto.createHash('sha256');
     const uiZipSha256 = uiZipHasher.update(Buffer.from(uiBytes)).digest('hex');
 
     const uiDir = path.join(this.fs.uisDir(this.holochainDataRoot), uiZipSha256);
-    // here we assume that if the uiDir exists, the hashes.json exists as well.
-    if (!fs.existsSync(path.join(uiDir, 'assets'))) {
-      fs.mkdirSync(uiDir, { recursive: true });
-
-      const hashes: Record<string, string> = {};
-      const zip = new AdmZip(Buffer.from(uiBytes));
-      zip.getEntries().forEach((entry) => {
-        // console.log(entry.entryName);
-        const hasher = crypto.createHash('sha256');
-        const data = entry.getData();
-        hasher.update(data);
-        const hash = hasher.digest('hex');
-        hashes[entry.entryName] = hash;
-      });
+    const assetsPath = path.join(uiDir, 'assets');
 
-      this.integrityChecker.storeToSignedJSON(path.join(uiDir, 'hashes.json'), hashes);
-      zip.extractAllTo(path.join(uiDir, 'assets'));
+    // here we assume that if the uiDir exists, the hashes.json exists as well.
+    if (!fs.existsSync(assetsPath)) {
+      this.createUiDirectory(uiDir, uiBytes);
     }
+
     if (icon) {
-      fs.writeFileSync(path.join(uiDir, 'icon'), icon);
-    } else {
-      const maybeIconPngPath = path.join(uiDir, 'assets', 'icon.png');
```

Comment:
It looks like this part got removed without replacement? I think we should still support storing and displaying an icon for apps installed from filesystem...

Reply:
If we enforce that it's a PNG file, we may also just store the plain icon.png file in this case.